Normalization differs for dependencies and dev-dependencies

I have a library that's tested by trybuild in a subproject (roughly this structure):

```
.
├── Cargo.lock
├── Cargo.toml
├── compile-fail
│   ├── Cargo.lock
│   ├── Cargo.toml
│   ├── src
│   │   └── lib.rs
│   └── tests
│       └── ui
│           ├── repro.rs
│           └── repro.stderr
└── src
    └── lib.rs
```

**src/lib.rs**

```rust
pub trait MyTrait {}

impl MyTrait for i32 {}
```

**compile-fail/src/lib.rs**

```rust
#[test]
fn ui() {
    let t = trybuild::TestCases::new();
    t.compile_fail("tests/ui/*.rs");
}
```

**compile-fail/tests/ui/repro.rs**

```rust
use some_library::MyTrait;

fn repro() -> impl MyTrait {}

fn main() {}
```

Notably, I include trybuild and my library as `dev-dependencies` since they are only being used for tests:

**compile-fail/Cargo.toml**

```toml
[package]
name = "alpha"
version = "0.1.0"
edition = "2024"

[dev-dependencies]
trybuild = "1.0"
some-library = { path = ".." }
```

This results in a non-normalized stderr:

**compile-fail/tests/ui/repro.stderr**

```
error[E0277]: the trait bound `(): MyTrait` is not satisfied
 --> tests/ui/repro.rs:3:15
  |
3 | fn repro() -> impl MyTrait {}
  |               ^^^^^^^^^^^^ the trait `MyTrait` is not implemented for `()`
  |
help: the trait `MyTrait` is implemented for `i32`
 --> /private/tmp/alpha/src/lib.rs:3:1
  |
3 | impl MyTrait for i32 {}
  | ^^^^^^^^^^^^^^^^^^^^
```

If I change to including it as `dependencies`, the path is normalized:

```
error[E0277]: the trait bound `(): MyTrait` is not satisfied
 --> tests/ui/repro.rs:3:15
  |
3 | fn repro() -> impl MyTrait {}
  |               ^^^^^^^^^^^^ the trait `MyTrait` is not implemented for `()`
  |
help: the trait `MyTrait` is implemented for `i32`
 --> $SOME_LIBRARY/src/lib.rs
  |
  | impl MyTrait for i32 {}
  | ^^^^^^^^^^^^^^^^^^^^
```

---

Ideally, both `dependencies` and `dev-dependencies` would be normalized the same way. Barring that, it'd be nice to note somewhere the intentional difference between the two. I actually didn't find any docs that mention that normalization occurred at all...

I've actually been using trybuild since mid-2019 and never realized that this feature existed / that I was missing it.

--- a/fuzz/fuzz_targets/normalize.rs
+++ b/fuzz/fuzz_targets/normalize.rs
@@ -9,7 +9,7 @@ mod directory;
 mod normalize;
 
 use crate::directory::Directory;
-use crate::normalize::Context;
+use crate::normalize::{Context, Normalization};
 use crate::run::PathDependency;
 use libfuzzer_sys::fuzz_target;
 use std::path::Path;
@@ -18,6 +18,7 @@ mod run {
     pub struct PathDependency {
         pub name: String,
         pub normalized_path: super::Directory,
+        pub normalization: super::Normalization,
     }
 }
 
@@ -34,6 +35,7 @@ fuzz_target!(|string: &str| {
         path_dependencies: &[PathDependency {
             name: String::from("diesel"),
             normalized_path: Directory::new("/home/user/documents/rust/diesel/diesel"),
+            normalization: Normalization::PathDependencies,
         }],
     };
     let _ = normalize::diagnostics(string, context);

--- a/src/normalize.rs
+++ b/src/normalize.rs
@@ -21,8 +21,8 @@ pub(crate) struct Context<'a> {
 
 macro_rules! normalizations {
     ($($name:ident,)*) => {
-        #[derive(PartialOrd, PartialEq, Copy, Clone)]
-        enum Normalization {
+        #[derive(PartialOrd, PartialEq, Copy, Clone, Debug)]
+        pub(crate) enum Normalization {
             $($name,)*
         }
 
@@ -65,6 +65,7 @@ normalizations! {
     HeadingNote,
     UnindentSuggestion,
     CustomRegistry,
+    MorePathDependencies,
     // New normalization steps are to be inserted here at the end so that any
     // snapshots saved before your normalization change remain passing.
 }
@@ -265,16 +266,19 @@ impl<'a> Filter<'a> {
             }
             if self.normalization >= PathDependencies && !other_crate {
                 for path_dep in self.context.path_dependencies {
-                    let path_dep_pat = path_dep
-                        .normalized_path
-                        .to_string_lossy()
-                        .to_ascii_lowercase()
-                        .replace('\\', "/");
-                    if let Some(i) = line_lower.find(&path_dep_pat) {
-                        let var = format!("${}", path_dep.name.to_uppercase().replace('-', "_"));
-                        line.replace_range(i..i + path_dep_pat.len() - 1, &var);
-                        other_crate = true;
-                        break;
+                    if self.normalization >= path_dep.normalization {
+                        let path_dep_pat = path_dep
+                            .normalized_path
+                            .to_string_lossy()
+                            .to_ascii_lowercase()
+                            .replace('\\', "/");
+                        if let Some(i) = line_lower.find(&path_dep_pat) {
+                            let var =
+                                format!("${}", path_dep.name.to_uppercase().replace('-', "_"));
+                            line.replace_range(i..i + path_dep_pat.len() - 1, &var);
+                            other_crate = true;
+                            break;
+                        }
                     }
                 }
             }

--- a/src/run.rs
+++ b/src/run.rs
@@ -7,7 +7,7 @@ use crate::expand::{ExpandedTest, expand_globs};
 use crate::flock::Lock;
 use crate::manifest::{Bin, Manifest, Name, Package, Workspace};
 use crate::message::{self, Fail, Warn};
-use crate::normalize::{self, Context, Variations};
+use crate::normalize::{self, Context, Normalization, Variations};
 use crate::path::CanonicalPath;
 use crate::{Expected, Runner, Test, features};
 use serde_derive::Deserialize;
@@ -39,6 +39,7 @@ pub(crate) struct Project {
 pub(crate) struct PathDependency {
     pub name: String,
     pub normalized_path: Directory,
+    pub normalization: Normalization,
 }
 
 struct Report {
@@ -131,25 +132,36 @@ impl Runner {
         let mut features = features::find();
 
         let mut path_dependencies = Vec::new();
-        let mut collect_path_dependencies = |dependencies: &Map<String, Dependency>| {
-            for (name, dep) in dependencies {
-                if let Some(path) = &dep.path
+        let mut collect_path_dependencies =
+            |dependencies: &Map<String, Dependency>, normalization: Normalization| {
+                for (name, dep) in dependencies {
+                    if let Some(path) = &dep.path
                     // Skip path dependencies coming from the workspace itself
                     && !packages.iter().any(|p| &p.name == name)
                     && let Ok(normalized_path) = path.canonicalize()
-                {
-                    path_dependencies.push(PathDependency {
-                        name: name.clone(),
-                        normalized_path,
-                    });
+                    {
+                        path_dependencies.push(PathDependency {
+                            name: name.clone(),
+                            normalized_path,
+                            normalization,
+                        });
+                    }
                 }
-            }
-        };
-        collect_path_dependencies(&source_manifest.dependencies);
-        collect_path_dependencies(&source_manifest.dev_dependencies);
+            };
+        collect_path_dependencies(
+            &source_manifest.dependencies,
+            Normalization::PathDependencies,
+        );
+        collect_path_dependencies(
+            &source_manifest.dev_dependencies,
+            Normalization::MorePathDependencies,
+        );
         for target in source_manifest.target.values() {
-            collect_path_dependencies(&target.dependencies);
-            collect_path_dependencies(&target.dev_dependencies);
+            collect_path_dependencies(&target.dependencies, Normalization::MorePathDependencies);
+            collect_path_dependencies(
+                &target.dev_dependencies,
+                Normalization::MorePathDependencies,
+            );
         }
 
         let crate_name = &source_manifest.package.name;

--- a/src/run.rs
+++ b/src/run.rs
@@ -130,22 +130,27 @@ impl Runner {
 
         let mut features = features::find();
 
-        let path_dependencies = source_manifest
-            .dependencies
-            .iter()
-            .filter_map(|(name, dep)| {
-                let path = dep.path.as_ref()?;
-                if packages.iter().any(|p| &p.name == name) {
+        let mut path_dependencies = Vec::new();
+        let mut collect_path_dependencies = |dependencies: &Map<String, Dependency>| {
+            for (name, dep) in dependencies {
+                if let Some(path) = &dep.path
                     // Skip path dependencies coming from the workspace itself
-                    None
-                } else {
-                    Some(PathDependency {
+                    && !packages.iter().any(|p| &p.name == name)
+                    && let Ok(normalized_path) = path.canonicalize()
+                {
+                    path_dependencies.push(PathDependency {
                         name: name.clone(),
-                        normalized_path: path.canonicalize().ok()?,
-                    })
+                        normalized_path,
+                    });
                 }
-            })
-            .collect();
+            }
+        };
+        collect_path_dependencies(&source_manifest.dependencies);
+        collect_path_dependencies(&source_manifest.dev_dependencies);
+        for target in source_manifest.target.values() {
+            collect_path_dependencies(&target.dependencies);
+            collect_path_dependencies(&target.dev_dependencies);
+        }
 
         let crate_name = &source_manifest.package.name;
         let project_dir = path!(target_dir / "tests" / "trybuild" / crate_name /);

--- a/src/tests.rs
+++ b/src/tests.rs
@@ -18,6 +18,7 @@ macro_rules! test_normalize {
                 path_dependencies: &[crate::run::PathDependency {
                     name: String::from("diesel"),
                     normalized_path: crate::directory::Directory::new("/home/user/documents/rust/diesel/diesel"),
+                    normalization: crate::normalize::Normalization::PathDependencies,
                 }],
             };
             let original = $original;

